File: scripts/self-train-loop.mjs
import { spawnSync } from "child_process";

function parseArgs(argv) {
  const out = {
    passthrough: [],
    trainer: "py",
    stopOnError: false,
    sleepMs: 2000,
    maxRuns: 0
  };
  const localOnly = new Set(["trainer", "stop-on-error", "sleep-ms", "max-runs"]);
  for (let i = 0; i < argv.length; i += 1) {
    const item = argv[i];
    if (!item.startsWith("--")) {
      out.passthrough.push(item);
      continue;
    }
    const key = item.slice(2);
    const next = argv[i + 1];
    const hasValue = Boolean(next) && !next.startsWith("--");
    if (!localOnly.has(key)) {
      out.passthrough.push(item);
      if (hasValue) {
        out.passthrough.push(next);
        i += 1;
      }
      continue;
    }

    if (key === "trainer") {
      if (hasValue) {
        out.trainer = String(next);
        i += 1;
      } else {
        out.trainer = "py";
      }
      continue;
    }
    if (key === "stop-on-error") {
      out.stopOnError = hasValue ? String(next).toLowerCase() === "true" : true;
      if (hasValue) i += 1;
      continue;
    }
    if (key === "sleep-ms") {
      const raw = hasValue ? Number(next) : 2000;
      if (hasValue) i += 1;
      out.sleepMs = Number.isFinite(raw) ? Math.max(0, Math.floor(raw)) : 2000;
      continue;
    }
    if (key === "max-runs") {
      const raw = hasValue ? Number(next) : 0;
      if (hasValue) i += 1;
      out.maxRuns = Number.isFinite(raw) ? Math.max(0, Math.floor(raw)) : 0;
    }
  }
  return out;
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  const trainer = String(args.trainer || "py").toLowerCase().trim();
  const stopOnError = Boolean(args.stopOnError);
  const sleepMs = Number.isFinite(args.sleepMs) ? Math.max(0, Math.floor(args.sleepMs)) : 2000;
  const maxRuns = Number.isFinite(args.maxRuns) ? Math.max(0, Math.floor(args.maxRuns)) : 0;
  const passthrough = Array.isArray(args.passthrough) ? args.passthrough : [];

  let stopping = false;
  process.on("SIGINT", () => {
    stopping = true;
    console.log("\nloop: stop requested (Ctrl+C), waiting current run to finish...");
  });
  process.on("SIGTERM", () => {
    stopping = true;
  });

  let round = 0;
  while (!stopping) {
    if (maxRuns > 0 && round >= maxRuns) break;
    round += 1;
    console.log(`loop: run #${round} (trainer=${trainer})`);
    const cmdArgs = ["scripts/self-train.mjs", "--trainer", trainer, ...passthrough];
    const res = spawnSync(process.execPath, cmdArgs, {
      cwd: process.cwd(),
      stdio: "inherit",
      shell: false
    });
    const code = Number(res.status ?? 1);
    if (code !== 0 && stopOnError) {
      process.exit(code);
    }
    if (stopping) break;
    if (sleepMs > 0) {
      await sleep(sleepMs);
    }
  }
  console.log("loop: stopped");
}

main().catch((err) => {
  console.error(err.message || String(err));
  process.exit(1);
});

File: scripts_py/self_train_loop.py
import signal
import subprocess
import sys
import time

from common_pipeline import parse_args, to_bool


def main():
    args = parse_args(sys.argv[1:])
    trainer = str(args.get("trainer") or "py").strip().lower()
    stop_on_error = to_bool(args.get("stop-on-error"), False)
    sleep_ms = int(float(args.get("sleep-ms") or 2000))
    max_runs = int(float(args.get("max-runs") or 0))

    passthrough = []
    local = {"trainer", "stop-on-error", "sleep-ms", "max-runs"}
    raw = sys.argv[1:]
    i = 0
    while i < len(raw):
      token = raw[i]
      if token.startswith("--"):
        key = token[2:]
        nxt = raw[i + 1] if i + 1 < len(raw) else None
        has_value = bool(nxt) and not nxt.startswith("--")
        if key in local:
          i += 2 if has_value else 1
          continue
        passthrough.append(token)
        if has_value:
          passthrough.append(nxt)
          i += 2
        else:
          i += 1
      else:
        passthrough.append(token)
        i += 1

    stopping = {"v": False}

    def _stop_handler(_sig, _frame):
        stopping["v"] = True
        print("\nloop: stop requested (Ctrl+C), waiting current run to finish...")

    signal.signal(signal.SIGINT, _stop_handler)
    signal.signal(signal.SIGTERM, _stop_handler)

    round_no = 0
    while not stopping["v"]:
        if max_runs > 0 and round_no >= max_runs:
            break
        round_no += 1
        print(f"loop: run #{round_no} (trainer={trainer})")
        cmd = ["python", "scripts_py/self_train.py", "--trainer", trainer] + passthrough
        res = subprocess.run(cmd, shell=False)
        code = int(res.returncode or 0)
        if code != 0 and stop_on_error:
            raise SystemExit(code)
        if stopping["v"]:
            break
        if sleep_ms > 0:
            time.sleep(sleep_ms / 1000.0)
    print("loop: stopped")


if __name__ == "__main__":
    main()
